### ISSUE TITLE
Prover fail to run

### ISSUE THREAD

**Issue by hussein-aitlahcen** [risc0/risc0#358]
Hi, while trying to wrap [our CosmWasm VM](https://github.com/ComposableFi/cosmwasm-vm) I encountered the following issue:
```
warning: `starter` (bin "starter") generated 1 warning
    Finished release [optimized] target(s) in 5m 35s
     Running `target/release/starter`
Reading prover.
Generating receipt...
thread 'main' panicked at 'impossible: eqz failed at: cirgen/components/u32.cpp:266', starter/src/main.rs:9:32
```

**Comment by SchmErik** [comment]
Hi @hussein-aitlahcen which version or commit ID are you using?

**Comment by hussein-aitlahcen** [comment]
@SchmErik the version 0.12

```toml
[[package]]
name = "risc0-zkp"
version = "0.12.0"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "a784efbf2967168ccdc93ffff150bfe9b167234735dea7c0a85ee7185dd4a6a4"
dependencies = [
 "anyhow",
 "bytemuck",
 "log",
 "ndarray",
 "paste",
 "rand",
 "rand_core",
 "rayon",
 "risc0-zeroio",
 "serde",
 "sha2",
 "tracing",
]

[[package]]
name = "risc0-zkvm"
version = "0.12.0"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "71e29d2c652dc7f7de1e82a7d4a95670b853f64ac54366c35130566860b51aea"
dependencies = [
 "anyhow",
 "bytemuck",
 "cfg-if",
 "elf",
 "lazy-regex",
 "log",
 "num-derive",
 "num-traits",
 "rand",
 "rayon",
 "risc0-circuit-rv32im",
 "risc0-zeroio",
 "risc0-zkp",
 "risc0-zkvm-platform",
 "serde",
 "sha2",
 "tracing",
]

[[package]]
name = "risc0-zkvm-platform"
version = "0.12.0"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "5ae26c13042eaf4d45655883b01c778aeebed68b8d0f44970d08e111074654d9"
```

**Comment by tzerrell** [comment]
I wonder if this has the same root cause as https://github.com/risc0/risc0-rust-examples/issues/47

**Comment by flaub** [comment]
> I wonder if this has the same root cause as [risc0/risc0-rust-examples#47](https://github.com/risc0/risc0-rust-examples/issues/47)

This is a different `eqz` failure: it's not memory related, it's related to the constraints for a u32.

**Comment by flaub** [comment]
This seems related to the same problem that we've found with tinygo. @jbruestle mentioned that this was a bug in the circuit that he was hoping to address soon.

**Comment by nategraf** [comment]
I believe I may have hit the same issue with the following branch of the SHA-256 example: https://github.com/risc0/risc0-rust-examples/tree/victor/pbkdf2-example

```bash
cd sha
RUST_BACKTRACE=full cargo run --release -- abc
```

```
thread 'main' panicked at 'Code should be provable: eqz failed at: cirgen/components/u32.cpp:268

Stack backtrace:
   0: std::backtrace_rs::backtrace::libunwind::trace
             at /rustc/0da281b6068a7d889ae89a9bd8991284cc9b7535/library/std/src/../../backtrace/src/backtrace/mod.rs:66:5
   1: std::backtrace_rs::backtrace::trace_unsynchronized
             at /rustc/0da281b6068a7d889ae89a9bd8991284cc9b7535/library/std/src/../../backtrace/src/backtrace/mod.rs:66:5
   2: std::backtrace::Backtrace::create
             at /rustc/0da281b6068a7d889ae89a9bd8991284cc9b7535/library/std/src/backtrace.rs:332:13
   3: anyhow::kind::Adhoc::new
   4: risc0_circuit_rv32im::ffi::call_step
   5: risc0_zkvm::prove::exec::RV32Executor<H>::run::{{closure}}
   6: risc0_zkvm::prove::loader::Loader::load
   7: risc0_zkvm::prove::exec::RV32Executor<H>::run
   8: risc0_zkvm::prove::Prover::run_with_hal
   9: risc0_zkvm::prove::Prover::run
  10: sha_example::main
  11: std::sys_common::backtrace::__rust_begin_short_backtrace
  12: std::rt::lang_start::{{closure}}
  13: core::ops::function::impls::<impl core::ops::function::FnOnce<A> for &F>::call_once
             at /rustc/0da281b6068a7d889ae89a9bd8991284cc9b7535/library/core/src/ops/function.rs:286:13
  14: std::panicking::try::do_call
             at /rustc/0da281b6068a7d889ae89a9bd8991284cc9b7535/library/std/src/panicking.rs:483:40
  15: std::panicking::try
             at /rustc/0da281b6068a7d889ae89a9bd8991284cc9b7535/library/std/src/panicking.rs:447:19
  16: std::panic::catch_unwind
             at /rustc/0da281b6068a7d889ae89a9bd8991284cc9b7535/library/std/src/panic.rs:137:14
  17: std::rt::lang_start_internal::{{closure}}
             at /rustc/0da281b6068a7d889ae89a9bd8991284cc9b7535/library/std/src/rt.rs:148:48
  18: std::panicking::try::do_call
             at /rustc/0da281b6068a7d889ae89a9bd8991284cc9b7535/library/std/src/panicking.rs:483:40
  19: std::panicking::try
             at /rustc/0da281b6068a7d889ae89a9bd8991284cc9b7535/library/std/src/panicking.rs:447:19
  20: std::panic::catch_unwind
             at /rustc/0da281b6068a7d889ae89a9bd8991284cc9b7535/library/std/src/panic.rs:137:14
  21: std::rt::lang_start_internal
             at /rustc/0da281b6068a7d889ae89a9bd8991284cc9b7535/library/std/src/rt.rs:148:20
  22: _main', host/src/main.rs:30:18
stack backtrace:
   0:        0x100f007e8 - std::backtrace_rs::backtrace::libunwind::trace::ha887c289731dd745
                               at /rustc/0da281b6068a7d889ae89a9bd8991284cc9b7535/library/std/src/../../backtrace/src/backtrace/mod.rs:66:5
   1:        0x100f007e8 - std::backtrace_rs::backtrace::trace_unsynchronized::heb3a46a9135817b5
                               at /rustc/0da281b6068a7d889ae89a9bd8991284cc9b7535/library/std/src/../../backtrace/src/backtrace/mod.rs:66:5
   2:        0x100f007e8 - std::sys_common::backtrace::_print_fmt::h80bf6884a559cc5b
                               at /rustc/0da281b6068a7d889ae89a9bd8991284cc9b7535/library/std/src/sys_common/backtrace.rs:65:5
   3:        0x100f007e8 - <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt::h304508731a5e5566
                               at /rustc/0da281b6068a7d889ae89a9bd8991284cc9b7535/library/std/src/sys_common/backtrace.rs:44:22
   4:        0x100f14ec0 - core::fmt::write::h927edcb79640dd65
                               at /rustc/0da281b6068a7d889ae89a9bd8991284cc9b7535/library/core/src/fmt/mod.rs:1209:17
   5:        0x100efe504 - std::io::Write::write_fmt::hed6f0d49e5386442
                               at /rustc/0da281b6068a7d889ae89a9bd8991284cc9b7535/library/std/src/io/mod.rs:1682:15
   6:        0x100f005fc - std::sys_common::backtrace::_print::he81c7da466074a18
                               at /rustc/0da281b6068a7d889ae89a9bd8991284cc9b7535/library/std/src/sys_common/backtrace.rs:47:5
   7:        0x100f005fc - std::sys_common::backtrace::print::ha0edf7f561957635
                               at /rustc/0da281b6068a7d889ae89a9bd8991284cc9b7535/library/std/src/sys_common/backtrace.rs:34:9
   8:        0x100f01e3c - std::panicking::default_hook::{{closure}}::hb3f1452a7b33b964
                               at /rustc/0da281b6068a7d889ae89a9bd8991284cc9b7535/library/std/src/panicking.rs:267:22
   9:        0x100f01b94 - std::panicking::default_hook::hcd4d8481747fb8a7
                               at /rustc/0da281b6068a7d889ae89a9bd8991284cc9b7535/library/std/src/panicking.rs:286:9
  10:        0x100f02478 - std::panicking::rust_panic_with_hook::h23a732d47cd44a6b
                               at /rustc/0da281b6068a7d889ae89a9bd8991284cc9b7535/library/std/src/panicking.rs:688:13
  11:        0x100f02288 - std::panicking::begin_panic_handler::{{closure}}::h3e18958f999b54ff
                               at /rustc/0da281b6068a7d889ae89a9bd8991284cc9b7535/library/std/src/panicking.rs:579:13
  12:        0x100f00c50 - std::sys_common::backtrace::__rust_end_short_backtrace::h39ea6b426428ea57
                               at /rustc/0da281b6068a7d889ae89a9bd8991284cc9b7535/library/std/src/sys_common/backtrace.rs:137:18
  13:        0x100f01fdc - rust_begin_unwind
                               at /rustc/0da281b6068a7d889ae89a9bd8991284cc9b7535/library/std/src/panicking.rs:575:5
  14:        0x100f999a8 - core::panicking::panic_fmt::h25d05832117e0f1b
                               at /rustc/0da281b6068a7d889ae89a9bd8991284cc9b7535/library/core/src/panicking.rs:65:14
  15:        0x100f99b48 - core::result::unwrap_failed::h9ae0a9449d7accc7
                               at /rustc/0da281b6068a7d889ae89a9bd8991284cc9b7535/library/core/src/result.rs:1791:5
  16:        0x100b47be0 - sha_example::main::ha6b72249d0b59a8e
  17:        0x100b3ac5c - std::sys_common::backtrace::__rust_begin_short_backtrace::h3d3ba6a4e567bb5e
  18:        0x100b3dde0 - std::rt::lang_start::{{closure}}::h894471d9ca131583
  19:        0x100ef9cbc - core::ops::function::impls::<impl core::ops::function::FnOnce<A> for &F>::call_once::h531184c45f180d10
                               at /rustc/0da281b6068a7d889ae89a9bd8991284cc9b7535/library/core/src/ops/function.rs:286:13
  20:        0x100ef9cbc - std::panicking::try::do_call::h372ce4d88372acd2
                               at /rustc/0da281b6068a7d889ae89a9bd8991284cc9b7535/library/std/src/panicking.rs:483:40
  21:        0x100ef9cbc - std::panicking::try::h5e0985787a1bd31b
                               at /rustc/0da281b6068a7d889ae89a9bd8991284cc9b7535/library/std/src/panicking.rs:447:19
  22:        0x100ef9cbc - std::panic::catch_unwind::h1b14e53c40992956
                               at /rustc/0da281b6068a7d889ae89a9bd8991284cc9b7535/library/std/src/panic.rs:137:14
  23:        0x100ef9cbc - std::rt::lang_start_internal::{{closure}}::h26e9003674d8c8c8
                               at /rustc/0da281b6068a7d889ae89a9bd8991284cc9b7535/library/std/src/rt.rs:148:48
  24:        0x100ef9cbc - std::panicking::try::do_call::ha500ce02b48e185e
                               at /rustc/0da281b6068a7d889ae89a9bd8991284cc9b7535/library/std/src/panicking.rs:483:40
  25:        0x100ef9cbc - std::panicking::try::h996a42d06d43e608
                               at /rustc/0da281b6068a7d889ae89a9bd8991284cc9b7535/library/std/src/panicking.rs:447:19
  26:        0x100ef9cbc - std::panic::catch_unwind::h3aa5334360ffe48e
                               at /rustc/0da281b6068a7d889ae89a9bd8991284cc9b7535/library/std/src/panic.rs:137:14
  27:        0x100ef9cbc - std::rt::lang_start_internal::heaf0a56d94091d88
                               at /rustc/0da281b6068a7d889ae89a9bd8991284cc9b7535/library/std/src/rt.rs:148:20
```

**Comment by flaub** [comment]
This might be fixed in: https://github.com/risc0/risc0/pull/359

**Comment by flaub** [comment]
@nategraf @hussein-aitlahcen I believe this is fixed in #359, please re-open this if you still see a problem.
